hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/utils/Pipelines.java

```diff
@@ -330,17 +330,23 @@ public static DataStream<Object> hoodieStreamWrite(Configuration conf, int defau
           .setParallelism(conf.getInteger(FlinkOptions.WRITE_TASKS));
     } else {
       WriteOperatorFactory<HoodieRecord> operatorFactory = StreamWriteOperator.getFactory(conf);
-      return dataStream
-          // Key-by record key, to avoid multiple subtasks write to a bucket at the same time
-          .keyBy(HoodieRecord::getRecordKey)
-          .transform(
-              "bucket_assigner",
-              TypeInformation.of(HoodieRecord.class),
-              new KeyedProcessOperator<>(new BucketAssignFunction<>(conf)))
-          .uid("uid_bucket_assigner_" + conf.getString(FlinkOptions.TABLE_NAME))
-          .setParallelism(conf.getOptional(FlinkOptions.BUCKET_ASSIGN_TASKS).orElse(defaultParallelism))
-          // shuffle by fileId(bucket id)
-          .keyBy(record -> record.getCurrentLocation().getFileId())
+
+      DataStream<HoodieRecord> bucketDataStream = dataStream
+              // Key-by record key, to avoid multiple subtasks write to a bucket at the same time
+              .keyBy(HoodieRecord::getRecordKey)
+              .transform(
+                      "bucket_assigner",
+                      TypeInformation.of(HoodieRecord.class),
+                      new KeyedProcessOperator<>(new BucketAssignFunction<>(conf)))
+              .uid("uid_bucket_assigner_" + conf.getString(FlinkOptions.TABLE_NAME))
+              .setParallelism(conf.getOptional(FlinkOptions.BUCKET_ASSIGN_TASKS).orElse(defaultParallelism));
+
+      bucketDataStream = conf.getOptional(FlinkOptions.BUCKET_ASSIGN_TASKS).orElse(defaultParallelism) ==
+              conf.getInteger(FlinkOptions.WRITE_TASKS) ? bucketDataStream : bucketDataStream
+              // shuffle by fileId(bucket id)
+              .keyBy(record -> record.getCurrentLocation().getFileId());
```

Comment:
> 

Thanks for your review, actually, the bucket data stream is a keyed stream that repartition by hoodie record key.
If the bucket assgin opeator 's task num equals the write operator's, we don't need the repartition which repartition by location.

Reply:
There will not be the differrent tasks that hold the same fileid. Cause the bucket assign operator task get the keyed stream that repatitioned by hoodie key which means that every bucket assign operator task hold the unique file id, and wirte operator task : bucket assgin operator task is 1 : 1 which is the preconditions of the improvement.

Reply:
Thanks, I see what you mean, when repartition, each task has decided to write only one fileid.

Reply:
Actually, I think we can build a map-only data stream when we don't need update or delete. I can create a new issue to discuss this for avoid confict.

Reply:
I checked the code. Before entering `bucket_assigner`, there is a `index_bootstrap` to initialize the `location` of each key, after repartition, different locations will enter the same task, then assign locations to different keys, So I think it is necessary to repartition according to fileid here.

Reply:
`bucket index` does not need to be repartitioned by `fileId`.

Reply:
@LinMingQiang  As the precondition, the bucket_assginer keep that the same bucket will not be handled in diffrent task.
if 
> conf.getOptional(FlinkOptions.BUCKET_ASSIGN_TASKS).orElse(defaultParallelism) ==
              conf.getInteger(FlinkOptions.WRITE_TASKS)

I think write_operator will keep that the same file id will not be handled by diffrent write task. That is why we try to  repartion according to file id.

Reply:
1. one key will be sended to the same bucket assginer task always
2. one bucket / fileid will be generated by the same bucket assginer task always
3. if bucket assginer task nums == write opeartor task nums : one bucket / fileid will be sended to the same write opeator task.

Reply:
@LinMingQiang So how do you look?

Reply:
i have no idea, This is different from what I understand, You can ask Danny on the DingDing.

Reply:
> i have no idea, This is different from what I understand, You can ask Danny on the DingDing.

@LinMingQiang THX, CC @danny0405

Reply:
I think it is a valid change but how much gains can we get here ?

Reply:
> I think it is a valid change but how much gains can we get here ?

As the result of our test, we can simply resolve the data skew problem. Also, we can reduce the network overhead if we chain the opeators. Above all, we can gain more resource usage and throughput.

Reply:
Why the data skew can happen ? The fileId is shuffled right ?

Reply:
> Then why not try to promote the hash algorithm, say the murmur hash ?

The hash algorithm what flink use is the murmur hash. 
![image](https://user-images.githubusercontent.com/34703664/177324565-7c8aea9e-fe6f-4ec6-98b1-f7dad7bce232.png)

The point is that the bucket num is small that equals to the downstream task num. 
See this example code:		
      _int bucketNum = 10;
		int taskNum = 10;
		int [] map = new int[taskNum];
		for (int i = 0; i < bucketNum ; i++) {
			map[Math.abs(murmurHash(UUID.randomUUID().hashCode())) % taskNum] += 1;
		}
		for (int num:
			map) {
			System.out.println(num);
		}_


if we just generate 10 buckets for 10 tasks, we can get like this:
![image](https://user-images.githubusercontent.com/34703664/177328816-6c13956c-d85e-4ba1-bcf8-cc27c7155b52.png)

if we generate more buckets for 10 tasks, it seems like more balance:
![image](https://user-images.githubusercontent.com/34703664/177328672-c6b56721-97f7-49ee-8246-0e345fc8d564.png)

Reply:
But we should figure out more random algorithm here instead of just fix the case when bucket assign and write task have the same parallelism, how about the bucket assign has parallelism 4 and write task has parallelism 6 here ?

Reply:
> But we should figure out more random algorithm here instead of just fix the case when bucket assign and write task have the same parallelism, how about the bucket assign has parallelism 4 and write task has parallelism 6 here ?

@danny0405 Yeah, I approval this. We do need to resolve the problem completely. I will think about a better idea later. 
But this pr can help resolve the network overhead and the data skew in some case. I think it should be a improvment not bug fix.

Reply:
It is a fix but the limitation is too strict here, so not that valid from my side.

Reply:
> It is a fix but the limitation is too strict here, so not that valid from my side.

Okay, I get you. Thanks for your reply whatever.